File: app/src/main/java/com/rosebay/odds/ui/mainOdds/MainActivity.java
package com.rosebay.odds.ui.mainOdds;

import android.os.Bundle;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.rosebay.odds.BuildConfig;
import com.rosebay.odds.Constants;
import com.rosebay.odds.OddsApplication;
import com.rosebay.odds.R;
import com.rosebay.odds.model.SingleOdd;
import com.rosebay.odds.ui.CreateSingleOddInterface;
import com.rosebay.odds.ui.OnUsernameSavedInterface;
import com.rosebay.odds.ui.createOdds.CreateOddsFragment;
import com.rosebay.odds.ui.disclaimer.DisclaimerFragment;
import com.rosebay.odds.ui.favoriteOdds.FavoriteOddsFragment;
import com.rosebay.odds.ui.myOdds.MyOddsFragment;
import com.rosebay.odds.ui.singleOdd.SingleOddFragment;
import com.rosebay.odds.util.SharedPreferencesClient;
import com.squareup.leakcanary.RefWatcher;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity implements OnUsernameSavedInterface, CreateSingleOddInterface {

    @Inject
    SharedPreferencesClient usernamePreferencesClient;

    @BindView(R.id.home_menu)
    BottomNavigationView mBottomMenu;
    @BindView(R.id.adView)
    AdView mAdView;

    long backPressed;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        OddsApplication.getAppComponent().inject(this);
        MobileAds.initialize(this, BuildConfig.AdMobID);
        AdRequest request = new AdRequest.Builder()
                .build();
        mAdView.loadAd(request);
        loadInitialView(usernamePreferencesClient.getUsername(getResources().getString(R.string.username)));
        mBottomMenu.setOnNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.home:
                    getMainOddsFragment();
                    return true;

                case R.id.createOdds:
                    getCreateOddsFragment();
                    return true;

                case R.id.favOdds:
                    getFavoriteOddsFragment();
                    return true;

                case R.id.myOdds:
                    getMyOddsFragment();
                    return true;

                default:
                    return true;
            }
        });
    }

    public void loadInitialView(String username) {
        if (username != null) {
            getMainOddsFragment();
            mBottomMenu.setVisibility(View.VISIBLE);
        } else {
            mBottomMenu.setVisibility(View.INVISIBLE);
            showDisclaimerDialog();
        }
    }

    @Override
    public void onUsernameSaved() {
        mBottomMenu.setVisibility(View.VISIBLE);
        getMainOddsFragment();
    }


    public void showDisclaimerDialog() {
        DisclaimerFragment fragment = DisclaimerFragment.Companion.newInstance();
        createFragment(fragment);
    }

    public void createFragment(Fragment fragment) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction().addToBackStack(null);
        transaction.replace(R.id.fragment_container, fragment);
        transaction.commit();
    }

    @Override
    public void getSingleOddsFragment(SingleOdd singleOdd) {
        SingleOddFragment fragment = SingleOddFragment.newInstance();
        Bundle args = new Bundle();
        args.putSerializable(Constants.SINGLE_ODD_KEY, singleOdd);
        fragment.setArguments(args);
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.add(R.id.fragment_container, fragment);
        fragmentTransaction.addToBackStack(fragment.getTag());
        fragmentTransaction.commit();
    }

    public void getMainOddsFragment() {
        MainOddsFragment fragment = MainOddsFragment.newInstance();
        getSupportFragmentManager().popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        createFragment(fragment);
    }

    public void getMyOddsFragment() {
        MyOddsFragment fragment = MyOddsFragment.Companion.newInstance();
        createFragment(fragment);
    }

    public void getCreateOddsFragment() {
        CreateOddsFragment fragment = CreateOddsFragment.newInstance();
        createFragment(fragment);
    }

    public void getFavoriteOddsFragment() {
        FavoriteOddsFragment fragment = FavoriteOddsFragment.newInstance();
        createFragment(fragment);
    }

    @Override
    public void onBackPressed() {
        if ((backPressed + 1000 > System.currentTimeMillis()) && (getSupportFragmentManager().getBackStackEntryCount() == 1)) {
            moveTaskToBack(true);
        } else if (getSupportFragmentManager().getBackStackEntryCount() == 1){
            Snackbar.make(mBottomMenu, R.string.back_btn_msg, Snackbar.LENGTH_SHORT).show();
        } else {
            mBottomMenu.getMenu().getItem(0).setChecked(true);
            getMainOddsFragment();
        }
        backPressed = System.currentTimeMillis();
    }

    @Override
    protected void onDestroy() {
        if (mAdView != null) {
            mAdView.destroy();
        }
        super.onDestroy();
            RefWatcher refWatcher = OddsApplication.getRefWatcher(getApplication());
            refWatcher.watch(this);
    }

}

File: app/src/main/java/com/rosebay/odds/ui/mainOdds/MainActivity.kt
package com.rosebay.odds.ui.mainOdds

import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import butterknife.BindView
import butterknife.ButterKnife
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.MobileAds
import com.rosebay.odds.BuildConfig
import com.rosebay.odds.OddsApplication
import com.rosebay.odds.R
import com.rosebay.odds.model.SingleOdd
import com.rosebay.odds.ui.CreateSingleOddInterface
import com.rosebay.odds.ui.OnUsernameSavedInterface
import com.rosebay.odds.ui.singleOdd.SingleOddFragment
import com.rosebay.odds.util.Constants
import com.rosebay.odds.util.FragmentFactoryInt
import com.rosebay.odds.util.SharedPreferencesClient
import javax.inject.Inject


class MainActivity : AppCompatActivity(), CreateSingleOddInterface, OnUsernameSavedInterface {

    @Inject
    lateinit var usernamePreferencesClient: SharedPreferencesClient
    @Inject
    lateinit var fragmentFactory: FragmentFactoryInt

    @BindView(R.id.home_menu)
    lateinit var mBottomMenu: BottomNavigationView
    @BindView(R.id.adView)
    lateinit var mAdView: AdView

    private var backPressed: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        OddsApplication.appComponent.inject(this)
        setManager(supportFragmentManager)
        MobileAds.initialize(this, BuildConfig.AdMobID)
        val request = AdRequest.Builder()
                .build()
        mAdView.loadAd(request)
        loadInitialView()
        mBottomMenu.setOnNavigationItemSelectedListener { navListener(it)}
    }

    private fun navListener(item: MenuItem) : Boolean {
        when (item.itemId) {
            R.id.home -> fragmentFactory.getMainOddsFragment()
            R.id.createOdds -> fragmentFactory.getCreateOddsFragment()
            R.id.favOdds -> fragmentFactory.getFavoriteOddsFragment()
            R.id.myOdds -> fragmentFactory.getMyOddsFragment()
            else -> return false
        }
        return true
    }

    fun loadInitialView() {
        if (!usernamePreferencesClient.getUsername(resources.getString(R.string.username)).isNullOrEmpty()) {
            fragmentFactory.getMainOddsFragment()
            mBottomMenu.visibility = View.VISIBLE
        } else {
            mBottomMenu.visibility = View.INVISIBLE
            fragmentFactory.getDisclaimerFragment()
        }
    }

    override fun onUsernameSaved() {
        mBottomMenu.visibility = View.VISIBLE
        fragmentFactory.getMainOddsFragment()
    }

    override fun getSingleOddsFragment(singleOdd: SingleOdd) {
        val fragment = SingleOddFragment.newInstance()
        val args = Bundle()
        args.putSerializable(Constants.SINGLE_ODD_KEY, singleOdd)
        args.putString(Constants.USERNAME, usernamePreferencesClient.getUsername(resources.getString(R.string.username)))
        fragment.arguments = args
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.add(R.id.fragment_container, fragment)
        fragmentTransaction.addToBackStack(fragment.tag)
        fragmentTransaction.commit()
    }

    override fun onBackPressed() {
        if (backPressed + 1000 > System.currentTimeMillis() && supportFragmentManager.backStackEntryCount == 1) {
            moveTaskToBack(true)
        } else if (supportFragmentManager.backStackEntryCount == 1) {
            Snackbar.make(mBottomMenu, R.string.back_btn_msg, Snackbar.LENGTH_SHORT).show()
        } else {
            mBottomMenu.menu.getItem(0).isChecked = true
            fragmentFactory.getMainOddsFragment()
        }
        backPressed = System.currentTimeMillis()
    }

    override fun onDestroy() {
        if (mAdView != null) {
            mAdView.destroy()
        }
        super.onDestroy()
        val refWatcher = OddsApplication.getRefWatcher(this)
        refWatcher.watch(this)
    }

    companion object {

        private lateinit var supportFragmentManager : FragmentManager
        fun setManager(fragmentManager: FragmentManager) {
            supportFragmentManager = fragmentManager
        }

        fun create(fragment: Fragment) {
            if (fragment is com.rosebay.odds.ui.mainOdds.MainOddsFragment) supportFragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
            val transaction = supportFragmentManager.beginTransaction().addToBackStack(null)
            transaction.replace(R.id.fragment_container, fragment)
            transaction.commit()
        }
    }

}

